### PR TITLE
feat: Display remote workers in webserver dashboard

### LINKED ISSUE #782: Display remote workers in webserver dashboard
- [ ] Integrate graphql client for webserver schema: https://github.com/TabbyML/tabby/blob/main/ee/tabby-webserver/graphql/schema.graphql
- [ ] Display worker card for workers returned from graphql endpoint: https://github.com/TabbyML/tabby/blob/main/ee/tabby-ui/app/(dashboard)/page.tsx#L101
- [ ] When there's at least one chat worker is available, display playground button: https://github.com/TabbyML/tabby/blob/184ccc81e9b61c706604ab8d1e5f1bb63f4dca1d/ee/tabby-ui/components/header.tsx#L33C15-L33C15

### PR DESCRIPTION
**Pull request by liangfung**
Close #782

### CODE REVIEW

File: ee/tabby-ui/lib/tabby-gql-client.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { GraphQLClient, Variables, RequestOptions } from 'graphql-request'
+
+export const graphQLClient = new GraphQLClient(
+  `${process.env.NEXT_PUBLIC_TABBY_SERVER_URL ?? ''}/graphql`,
+  {
+    credentials: 'include',
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/tabby-gql-client.ts:6]: cors work after removing this line - please research and see what's the reason behind.

**Comment by wsxiaoys** [comment]
I encounter following warning when running `next dev`:

```
 ⚠ ./node_modules/node-fetch/lib/index.js
Module not found: Can't resolve 'encoding' in '/Users/meng/Projects/tabby/ee/tabby-ui/node_modules/node-fetch/lib'

Import trace for requested module:
./node_modules/node-fetch/lib/index.js
./node_modules/cross-fetch/dist/node-ponyfill.js
./node_modules/graphql-request/build/esm/index.js
./lib/tabby-gql-client.ts
./lib/hooks/use-graphql.ts
./lib/hooks/use-remote-worker.ts
./app/(dashboard)/page.tsx
```

File: ee/tabby-ui/package.json [hunk 2]
```diff
@@ -11,7 +11,9 @@
     "preview": "next build && next start",
     "type-check": "tsc --noEmit",
     "format:write": "prettier --write \"{app,lib,components}/**/*.{ts,tsx,mdx}\" --cache",
-    "format:check": "prettier --check \"{app,lib,components}**/*.{ts,tsx,mdx}\" --cache"
+    "format:check": "prettier --check \"{app,lib,components}**/*.{ts,tsx,mdx}\" --cache",
+    "codegen": "graphql-codegen --config codegen.ts",
```
Comment (wsxiaoys) [inline ee/tabby-ui/package.json:15]: can this step being merged into `next build` step? usually there'll be some webpack config for this

File: ee/tabby-ui/lib/hooks/use-remote-worker.ts [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    if (!haveRemoteChatWorkers && healthInfo?.chat_model) {
+      _workers.push(transformHealthInfoToWorker(healthInfo, WorkerKind.Chat))
+    }
+    return groupBy(_workers, 'kind')
+  }, [healthInfo, workers])
+
+  return groupedWorkers
+}
+
+export { useRemoteWorkers, useMergedWorkers }
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/hooks/use-remote-worker.ts:60]: Guess you can export only `useMergedWorkers`? or just call it `useWorkers`

File: ee/tabby-ui/package.json (hunk 2 above)
Comment (liangfung) [reply ee/tabby-ui/package.json:18]: maybe using a prebuild script is more convenient.

File: ee/tabby-ui/lib/hooks/use-remote-worker.ts (hunk 3 above)
Comment (liangfung) [reply ee/tabby-ui/lib/hooks/use-remote-worker.ts]: done, just call it `useWorkers`

File: ee/tabby-ui/lib/hooks/use-worker.ts
Comment (wsxiaoys) [inline ee/tabby-ui/lib/hooks/use-worker.ts:1]: nit: file name should be use-workers

File: ee/tabby-ui/lib/hooks/use-worker.ts [hunk 4]
```diff
@@ -32,16 +26,16 @@ function transformHealthInfoToWorker(
   }
 }
 
-function useMergedWorkers(healthInfo: HealthInfo | undefined) {
-  const { data } = useRemoteWorkers()
+function useWorkers(healthInfo?: HealthInfo) {
```
Comment (wsxiaoys) [inline ee/tabby-ui/lib/hooks/use-worker.ts:29]: just call useHealthInfo here?

**Comment by liangfung** [comment]
> I encounter following warning when running `next dev`:
> 
> ```
>  ⚠ ./node_modules/node-fetch/lib/index.js
> Module not found: Can't resolve 'encoding' in '/Users/meng/Projects/tabby/ee/tabby-ui/node_modules/node-fetch/lib'
> 
> Import trace for requested module:
> ./node_modules/node-fetch/lib/index.js
> ./node_modules/cross-fetch/dist/node-ponyfill.js
> ./node_modules/graphql-request/build/esm/index.js
> ./lib/tabby-gql-client.ts
> ./lib/hooks/use-graphql.ts
> ./lib/hooks/use-remote-worker.ts
> ./app/(dashboard)/page.tsx
> ```

[https://github.com/jasonkuhrt/graphql-request/issues/448](https://github.com/jasonkuhrt/graphql-request/issues/448)
[https://github.com/jasonkuhrt/graphql-request/pull/597](https://github.com/jasonkuhrt/graphql-request/pull/597)
@wsxiaoys  The latest version of `graphql-request` still has the `cross-fetch` polyfill. The simple way to remove the harmless warning is installing a devDependency `encoding`. Or just remove `graphql-request`

**Comment by wsxiaoys** [comment]
> > I encounter following warning when running `next dev`:
> > ```
> >  ⚠ ./node_modules/node-fetch/lib/index.js
> > Module not found: Can't resolve 'encoding' in '/Users/meng/Projects/tabby/ee/tabby-ui/node_modules/node-fetch/lib'
> > 
> > Import trace for requested module:
> > ./node_modules/node-fetch/lib/index.js
> > ./node_modules/cross-fetch/dist/node-ponyfill.js
> > ./node_modules/graphql-request/build/esm/index.js
> > ./lib/tabby-gql-client.ts
> > ./lib/hooks/use-graphql.ts
> > ./lib/hooks/use-remote-worker.ts
> > ./app/(dashboard)/page.tsx
> > ```
> 
> [jasonkuhrt/graphql-request#448](https://github.com/jasonkuhrt/graphql-request/issues/448) [jasonkuhrt/graphql-request#597](https://github.com/jasonkuhrt/graphql-request/pull/597) @wsxiaoys The latest version of `graphql-request` still has the `cross-fetch` polyfill. The simple way to remove the harmless warning is installing a devDependency `encoding`. Or just remove `graphql-request`

I’m fine either way - as long as the warning can be silented.

**Pull request merged**